kv: tryAsyncAbort on correct transaction in TxnCoordSender

nvb opened the pull request:
Fixes #21620.
Fixes #21664.
Fixes #21733.
Fixes #21737.
Fixes #21738.
Fixes #21739.
Fixes #21761.
Fixes #21763.

This change fixes two bugs and a test flake that fell out of dc22c39. In that PR,
`TxnCoordSender` was re-scoped to the lifetime of a single `TxnID`, instead
of holding state for all ongoing transactions like it used to. Now when a transaction
aborts, `client.Txn` creates a new `TxnCoordSender` for the new transaction ID,
discarding the old one.

### bug: aborting incorrect transaction

Surprisingly, I was seeing instances where a `TxnCoordSender` was aborting
the wrong `Txn`. More specifically, a `TxnCoordSender` was immediately aborting
a brand new incarnation of a `Txn` instead of performing an async abort on
the old `Txn` that had seen the retryable error (and that it was bound to).
It looks like [the change on this line](https://github.com/cockroachdb/cockroach/commit/dc22c39cb98ebd3ddd8a021e74fb85526e9f16da#diff-cde6b3c94c7ddae4bb6d6a5b7bf5e8d9L958)
was the issue. The change deleted the `defer`, which was used to "avoiding a
race with the state update below." Adding the defer back in fixes the issue.

### flake: map TransactionReplayErrors into TxnPrevAttemptErrors

`client.Txn` allows multiple requests to be sent concurrently, even if
one contains a `BeginTxn` and the others don't. This is safe, but can
have subtle behavior if one of the other requests results in a txn abort
error. In this case, a `tryAsyncAbort` attempt due to the request that saw
the error may beat the original BeginTxn request to the transaction record.
This can result in a `TransactionReplayError`, which we previously allowed
to boil up to the client. This was always possible, but recent changes
to `TxnCoordSender` have made it much more likely because `TxnCoordSender`
is now more proactive about performing aborts.

We now catch this error and detect if it was meant for a previous Txn
incarnation. In this case, we replace it with a `TransactionReplayError`,
which is already handled at a higher level.

### bug: augmenting TxnCoordMeta of  incorrect transaction

distSQLReceiver calls into client.Txn to provide it with more
information about updates to its state made remotely. This change
makes sure that it does so before checking to see if the transaction
should be aborted, not after. Previously, if the transaction was
aborted then the AugmentTxnCoordMeta call would go to the wrong
transaction, in some cases setting its `Writing` flag and causing
it to skip its BeginTxn request.

nvb committed:
kv: bind Txn to TxnCoordSender immediately

This change fixes a bug introduced in dc22c39. In that PR, `TxnCoordSender`
was re-scoped to the lifetime of a single `TxnID`, instead of holding state
for all ongoing transactions. Now when a transaction aborts, `client.Txn`
creates a new `TxnCoordSender` for the new transaction ID, discarding the
old one.

Surprisingly, I was seeing instances where a `TxnCoordSender` was aborting
the wrong `Txn`. More specifically, a `TxnCoordSender` was immediately aborting
a brand new incarnation of a `Txn` instead of performing an async abort on
the old `Txn` that had seen the retryable error (and that it was bound to).
The issue was that we never actually "bound" a `TxnCoordSender` to its `Txn`
before calling `updateState` after the first response. This meant that
if the first response resulted in a transaction abort, a call to
`Txn.Update` would bind it to the new transaction, instead of the old.
To fix this, we now bind the `TxnCoordSender` immediately in its `Send`
method if it's not already set.

Release note: None

nvb committed:
roachpb: deprecate UntrackedTxnError

It was no longer in use and only made sense in the context of the old
TxnCoordSender.

Release note: None

nvb committed:
kv: return TxnPrevAttemptError for stale TransactionReplayErrors

client.Txn allows multiple requests to be sent concurrently, even if
one contains a BeginTxn and the others don't. This is safe, but can
have subtle behavior if one of the other requests results in a txn abort
error. In this case, a tryAsyncAbort attempt due to the request that saw
the error may beat the original BeginTxn request to the transaction record.
This can result in a TransactionReplayError, which we previously allowed
to boil up to the client. This was always possible, but recent changes
to TxnCoordSender have made it much more likely because TxnCoordSender
is now more proactive about performing aborts.

We now catch this error and detect if it was meant for a previous Txn
incarnation. In this case, we replace it with a TransactionReplayError,
which is already handled at a higher level.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21730)
<!-- Reviewable:end -->

andreimatei commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 935 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-L3d-qFgBNDCxTOSYOo1:-L3d-qFgBNDCxTOSYOo2:b-tu5rjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9884a71bee4effc42d86fa9de392a269ea04f232/pkg/kv/txn_coord_sender.go#L935)):*
> ```Go
> 			// restated transaction.
> 			if errTxnID != newTxn.ID {
> 				// defer(), avoiding a race with the state update below.
> ```

I don't understand how this works - doesn't deferring it make it always use the new/wrong txn id?
Even if it happens to work somehow, how about making `tryAsyncAbort()` a standalone function that takes in the txn and heartbeat channel and whatever else it needs explicitly, and calling it here with the id that we want? If that's feasible, I think it'd be a lot better.

---

*[pkg/roachpb/errors.proto, line 333 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-L3d0XhBRfxUAfgAn1H5:-L3d0XhBRfxUAfgAn1H6:bm340t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/140a43798111150ea08bd0cda9b31cbddfc49a08/pkg/roachpb/errors.proto#L333)):*
> ```protobuf
> 
> // DeprecatedUntrackedTxnError is deprecated.
> message DeprecatedUntrackedTxnError {
> ```

why are we keeping this sucker around? Can't we just reserve the field in the union?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 1 files at r1, 5 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 935 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-L3d-qFgBNDCxTOSYOo1:-L3d3I1wh8YQQZGdm51x:by9ayze) ([raw file](https://github.com/cockroachdb/cockroach/blob/9884a71bee4effc42d86fa9de392a269ea04f232/pkg/kv/txn_coord_sender.go#L935)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't understand how this works - doesn't deferring it make it always use the new/wrong txn id?
Even if it happens to work somehow, how about making `tryAsyncAbort()` a standalone function that takes in the txn and heartbeat channel and whatever else it needs explicitly, and calling it here with the id that we want? If that's feasible, I think it'd be a lot better.
</blockquote></details>

Yeah, the old `defer` passed in the specific transaction to be aborted. Also, using a `defer` in the middle of a function instead of a do/undo pair is weird. Since `tryAsyncAbort` is async, I don't see why it matters whether we start it now or at the end of this function. 

---

*[pkg/sql/session.go, line 774 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-L3d3xpzeez20Jq8c3lp:-L3d3xpzeez20Jq8c3lq:b-j3d5xa) ([raw file](https://github.com/cockroachdb/cockroach/blob/bae7465a3160d68594613cc709ba3f05c12ee40f/pkg/sql/session.go#L774)):*
> ```Go
> 					retryErr = t
> 				}
> 			case *roachpb.UntrackedTxnError:
> ```

Couldn't this still happen in mixed-version clusters? We can rename it, but I think we need to keep the uses of it around for a release cycle. (if not, we might as well just delete the proto and reserve the tag as Andrei suggested)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
It looks like the first commit here also fixes TestMonotonicInserts.

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 935 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-L3d-qFgBNDCxTOSYOo1:-L3d8J_TIjzu0KsASLHP:bt9jpgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9884a71bee4effc42d86fa9de392a269ea04f232/pkg/kv/txn_coord_sender.go#L935)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, the old `defer` passed in the specific transaction to be aborted. Also, using a `defer` in the middle of a function instead of a do/undo pair is weird. Since `tryAsyncAbort` is async, I don't see why it matters whether we start it now or at the end of this function. 
</blockquote></details>

The reason this was changed to no longer take a specific transaction is because `TxnCoordSender` should now only be scoped to a single transaction. The defer allowed us to ensure that the correct transaction was being bound to the `TxnCoordSender`.

I just replaced this defer with a better approach though. PTAL.

---

*[pkg/roachpb/errors.proto, line 333 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-L3d0XhBRfxUAfgAn1H5:-L3d7MUrqky4Si4YB1rD:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/140a43798111150ea08bd0cda9b31cbddfc49a08/pkg/roachpb/errors.proto#L333)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why are we keeping this sucker around? Can't we just reserve the field in the union?
</blockquote></details>

Good point. Done.

---

*[pkg/sql/session.go, line 774 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-L3d3xpzeez20Jq8c3lp:-L3d7L4lcmH1adoru-Qk:bxfq76w) ([raw file](https://github.com/cockroachdb/cockroach/blob/bae7465a3160d68594613cc709ba3f05c12ee40f/pkg/sql/session.go#L774)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Couldn't this still happen in mixed-version clusters? We can rename it, but I think we need to keep the uses of it around for a release cycle. (if not, we might as well just delete the proto and reserve the tag as Andrei suggested)
</blockquote></details>

I don't think so because it was only returned by `TxnCoordSender`, so the error should never be passed between nodes. I'm going to delete it, per Andrei's suggestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-:-L3dIM7fZd_PenFIIwQG:bhzfidl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Actually, no it doesn't. [This](https://github.com/cockroachdb/cockroach/pull/21759) does.

nvb committed:
distsql: AugmentTxnCoordMeta before aborting txn, not after

distSQLReceiver calls into client.Txn to provide it with more
information about updates to its state made remotely. This change
makes sure that it does so before checking to see if the transaction
should be aborted, not after. Previously, if the transaction was
aborted then the AugmentTxnCoordMeta call would go to the wrong
transaction, in some cases setting its `Writing` flag and causing
it to skip its BeginTxn request.

Release note: None

nvanbenschoten commented:
Ok, the final commit should now fix https://github.com/cockroachdb/cockroach/issues/21664.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21730#-:-L3e9hY_Vbdylj8fXP8Z:bnfp4nl)*
<!-- Sent from Reviewable.io -->
